keyboards/ymdk_sp64/matrix.c

```diff
@@ -0,0 +1,187 @@
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "matrix.h"
+#include <stdint.h>
+#include <stdbool.h>
+#include <avr/io.h>
+#include "wait.h"
+#include "action_layer.h"
+#include "print.h"
+#include "debug.h"
+#include "util.h"
+#include "ymdk_sp64.h"
+#include "debounce.h"
+
+#ifndef DEBOUNCE
+# define DEBOUNCE	5
+#endif
+
+
+/* matrix state(1:on, 0:off) */
+static matrix_row_t matrix[MATRIX_ROWS];
+
+// Debouncing: store for each key the number of scans until it's eligible to
+// change.  When scanning the matrix, ignore any changes in keys that have
+// already changed in the last DEBOUNCE scans.
+static matrix_row_t matrix_debouncing[MATRIX_ROWS];
+
+static void matrix_select_row(uint8_t row);
+
+#ifdef RIGHT_HALF
+static uint8_t mcp23018_reset_loop = 0;
+#endif
+
+void matrix_init(void)
+{
+  // all outputs for rows high
+  DDRB = 0xFF;
+  PORTB = 0xFF;
+  // all inputs for columns
+  DDRA = 0x00;
+  DDRC &= ~(0x111111<<2);
+  DDRD &= ~(1<<PIND7);
+  // all columns are pulled-up
+  PORTA = 0xFF;
+  PORTC |= (0b111111<<2);
+  PORTD |= (1<<PIND7);
+
+#ifdef RIGHT_HALF
+  // initialize row and col
+  mcp23018_status = init_mcp23018();
+#endif
+
+  // initialize matrix state: all keys off
+  for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+    matrix[row] = 0;
+    matrix_debouncing[row] = 0;
+  }
+
+
+  debounce_init(MATRIX_ROWS);
+  matrix_init_quantum();
+}
+
+/*static uint8_t bit_reverse(uint8_t x) {
+  x = ((x >> 1) & 0x55) | ((x << 1) & 0xaa);
+  x = ((x >> 2) & 0x33) | ((x << 2) & 0xcc);
+  x = ((x >> 4) & 0x0f) | ((x << 4) & 0xf0);
+  return x;
+}*/
+
+uint8_t matrix_scan(void)
+{
+#ifdef RIGHT_HALF
+  // Then the keyboard
+  if (mcp23018_status != I2C_STATUS_SUCCESS) {
+    if (++mcp23018_reset_loop == 0) {
+      // if (++mcp23018_reset_loop >= 1300) {
+      // since mcp23018_reset_loop is 8 bit - we'll try to reset once in 255 matrix scans
+      // this will be approx bit more frequent than once per second
+      print("trying to reset mcp23018\n");
+      mcp23018_status = init_mcp23018();
+      if (mcp23018_status) {
+        print("left side not responding\n");
+      } else {
+        print("left side attached\n");
+      }
+    }
+  }
+#endif
+
+  bool changed = false; 
+  
+  for (uint8_t row = 0; row < MATRIX_ROWS; row++)
+  {
+    matrix_row_t cols;
+
+    matrix_select_row(row);
+#ifndef RIGHT_HALF
+    _delay_us(5);
+#endif
+
+    cols = (
+      // cols 0..7, PORTA 0 -> 7
+      (~PINA) & 0xFF
+    );
+
+#ifdef RIGHT_HALF
+    uint8_t data = 0x7F;
+    // Receive the columns from right half
+    i2c_receive(I2C_ADDR_WRITE, &data, 1, MCP23018_I2C_TIMEOUT);
+    cols |= ((~(data) & 0x7F) << 7);
+#endif
+
+    if (matrix_debouncing[row] != cols) {
+      matrix_debouncing[row] = cols;
+      changed = true;
+    }
+  }
+  
+  debounce(matrix, matrix_debouncing, MATRIX_ROWS, changed);
```

Comment:
I might have this wrong. 
```suggestion
  debounce(matrix_debouncing, matrix, MATRIX_ROWS, changed);
```

Reply:
This was the smoking gun on why it wasn't working beforehand.